samza-core/src/main/java/org/apache/samza/zk/ZkLeaderElector.java

```diff
@@ -105,12 +105,12 @@ public void tryBecomeLeader(LeaderElectorListener leaderElectorListener) {
     LOG.info("Index = " + index + " Not eligible to be a leader yet!");
     String predecessor = children.get(index - 1);
     if (!predecessor.equals(currentSubscription)) {
-      if (currentSubscription != null) {
-
-        // callback in case if the previous node gets deleted (when previous processor dies)
-        if (previousProcessorChangeListener == null)
-          previousProcessorChangeListener =  new PreviousProcessorChangeListener(leaderElectorListener);
+      
+      // callback in case if the previous node gets deleted (when previous processor dies)
+      if (previousProcessorChangeListener == null)
+        previousProcessorChangeListener =  new PreviousProcessorChangeListener(leaderElectorListener);
```

Comment:
This check is incorrect, imo. If you create a new instance of the DataListener here, when you unsubscribe it is unsubscribing with a different reference.. This logic is very confusing. I have simplified it in SAMZA-1150. Please take a look.